// @sap/approuter/lib/backend-request/headers.js
'use strict';

var _ = require('lodash');
var url = require('url');
var uuid = require('uuid/v4');
var headerUtil = require('../utils/header-util');
var pathUtil = require('../utils/path-util');
var cookieUtils = require('../utils/cookie-utils');
var businessServiceUtils = require('../utils/business-service-utils');
var dynamicRoutingUtils = require('../utils/dynamic-routing-utils');

var HEADERS_TO_REMOVE = ['connection',
  'keep-alive',
  'public',
  'proxy-authenticate',
  'transfer-encoding',
  'upgrade',
  'x-approuter-authorization',
  'sec-websocket-key'];

exports.getHeaders = function (req, accessToken, destination) {
  var headers;
  var route;
  var tokens = {
    // In flow of Logout:
    // env destination => we get JWT token in tokens.accessToken as a String
    // destination service destination => we get destination auth token in tokens.accessToken as an Object
    accessToken: accessToken
  };
  if (req) {
    tokens.accessToken = tokens.accessToken || (req.session && req.session.user && req.session.user.token &&
      req.session.user.token.accessToken);
    destination = destination || req.internalUrl.destination;
    var servicePreserveHostHeader = req.internalUrl && req.internalUrl.route && req.internalUrl.route.service &&
      req.internalUrl.route.credentials && req.internalUrl.route.credentials.preserve_host_header;

    if ((destination && destination.preserveHostHeader) || servicePreserveHostHeader) {
      headers = _.omit(req.headers, HEADERS_TO_REMOVE);
    } else {
      headers = _.omit(req.headers, HEADERS_TO_REMOVE.concat(['host']));
    }
    if (destination && destination['sap-client']) {
      headers['sap-client'] = destination['sap-client'];
    }
    if (destination && destination.proxyType === 'OnPremise' && req.app.services && req.app.services['connectivity']) {
      tokens['connectivity'] = req.tenant && req.app.services['connectivity'][req.tenant] && req.app.services['connectivity'][req.tenant].token ?
        req.app.services['connectivity'][req.tenant].token : req.app.services['connectivity'].token;
    }

    addCorrelationIdHeader(headers, req);
    addXForwardingHeaders(headers, req, destination);
    removeSecurityHeaders(headers, req);
    if (req.headers['x-approuter-authorization'] && headers.cookie) {
      const backendName = cookieUtils.getBackendName(req);
      headers.cookie = cookieUtils.decryptCookies(headers.cookie, backendName);
    }
    headerUtil.updateSapPassport(headers);
    if (req.session && req.session.user && req.session.user.destinations
        && req.session.user.destinations[destination.name]
        && req.session.user.destinations[destination.name].authToken){
      tokens['authToken'] = req.session.user.destinations[destination.name].authToken;
    } else if (req.internalUrl && req.internalUrl.route
      && req.internalUrl.route.destination && req.internalUrl.route.authToken){
      tokens['authToken'] = req.internalUrl.route.authToken;
    }
    route = req.internalUrl ? req.internalUrl.route : null;
    if (route && route.service && req.app && req.app.services && req.app.services[route.service] && req.app.services[route.service].token){
      tokens[route.service] = req.app.services[route.service].token;
    }
    if (route && route.service === 'html5-apps-repo-rt') {
      let appKey = dynamicRoutingUtils.getApplicationKey(req);
      if (appKey && appKey.appPrefix) {
        addAppHostIdHeader(appKey.appPrefix, headers, req);
      }
    }
  }
  return addOauthHeader(headers || {}, tokens, route, destination);
};

function removeSecurityHeaders(headers, req) {
  if (!pathUtil.isPublicPath(req)) {
    delete headers['authorization'];
    if (pathUtil.isCsrfProtectionEnabled(req)) {
      delete headers['x-csrf-token'];
    }
  }
}

function addOauthHeader(headers, tokens, route, destination) {
  if (route && route.destination) { // Destination flow
    if (destination.proxyType && destination.proxyType === 'OnPremise') { // On premise destination flow - cloud connector
      headers['SAP-Connectivity-Authentication'] = 'Bearer ' + tokens.accessToken;
      headers['Proxy-Authorization'] = 'Bearer ' +  tokens['connectivity'].accessToken;
      if (destination.cloudConnectorLocationId) {
        headers['SAP-Connectivity-SCC-Location_ID'] = destination.cloudConnectorLocationId;
      }
    } else { // Cloud destination flow
      if (tokens && tokens.accessToken && destination.forwardAuthToken) {
        headers.authorization = 'Bearer ' + tokens.accessToken;
      }
    }
       // Common for on premise and cloud destination flows
       // Use tokens provided from destination service
    if (tokens['authToken']) {
      headers.authorization = tokens['authToken'].type + ' ' + tokens['authToken'].value;
    }
  } else if (route && route.service) {// Service flow
    if (businessServiceUtils.getGrantType(route.credentials) === 'client_credentials') {// html5 apps repo flow
      headers.authorization = tokens[route.service].tokenType + ' ' + tokens[route.service].accessToken;
    } else {
      headers.authorization = 'Bearer ' + tokens.accessToken;
    }
  } else {// WS or Logout
    if (tokens && tokens.accessToken) {
      //  In case of Logout:
      // env destination => we send JWT token in tokens.accessToken as a String
      // destination service destination => we send destination auth token in tokens.accessToken as an Object
      var tokenType = tokens.accessToken.type ? tokens.accessToken.type : 'Bearer';
      var tokenValue = tokens.accessToken.value ? tokens.accessToken.value : tokens.accessToken;
      headers.authorization = tokenType + ' ' + tokenValue;
    }
  }
  return headers;
}

function addCorrelationIdHeader(headers, req) {
  if (req.headers['x-correlationid']) {
    return;
  } else if (req.headers['x-request-id']) {
    headers['x-correlationid'] = req.headers['x-request-id'];
  } else if (req.headers['x-vcap-request-id']){
    headers['x-correlationid'] = req.headers['x-vcap-request-id'];
  } else {
    headers['x-correlationid'] = uuid();
  }
}

function addXForwardingHeaders(headers, req, destination) {
  if (destination && destination.setXForwardedHeaders === false){
    return;
  }
  if (req.headers.host) {
    headers['x-forwarded-host'] = req.headers['x-forwarded-host'] || req.headers.host;
  }
  if (!headers['x-forwarded-proto'] && req.protocol) {
    headers['x-forwarded-proto'] = req.headers['x-forwarded-proto'] || req.protocol;
  }
  if (!headers['x-forwarded-for']) {
    headers['x-forwarded-for'] = req.connection.remoteAddress;
  }
  if (!headers['x-forwarded-path']) {
    headers['x-forwarded-path'] = req.headers['x-forwarded-path'] || url.parse(req.url).pathname;
  }
}

function addAppHostIdHeader(businessServiceName, headers, req) {
  var businessServiceCredentials = businessServiceUtils.getCredentials(businessServiceName, true, req);
  if (businessServiceCredentials && businessServiceCredentials['html5-apps-repo'] && businessServiceCredentials['html5-apps-repo']['app_host_id']) {
    headers['x-app-host-id'] = businessServiceCredentials['html5-apps-repo']['app_host_id'];
  }
}
