// @sap/approuter/lib/utils/uaa-utils.js
'use strict';

var _ = require('lodash');
var url = require('url');
var logUtil = require('./logger');
var urlUtils = require('../utils/url-utils');

var tracer = logUtil.getTracer(__filename);

exports.resolveUaaConfig = resolveUaaConfig;
exports.retrieveTenantFromURL = retrieveTenantFromURL;

exports.getUaaConfig = function(req, cb) {
  var uaaOptions = req.extUaaConfigOptions || req.routerConfig.uaaConfig.options;
  if (req.app && req.app.approuter && req.app.approuter.resolveUaaConfig) {
    req.app.approuter.resolveUaaConfig(req, uaaOptions, cb);
  } else {
    resolveUaaConfig(req, uaaOptions, cb);
  }
};

function resolveUaaConfig(request, uaaOptions, callback) {
  if (uaaOptions.tenantmode !== 'shared') {
    callback(null, uaaOptions);
  } else {
    uaaOptions = _.clone(uaaOptions);
    var requestHost = urlUtils.getAppRouterHost(request);
    var tenantHostPattern = request.routerConfig.uaaConfig.tenantHostPattern;
    if (!tenantHostPattern){
      callback ('TENANT_HOST_PATTERN is not defined for tenant-mode shared');
    } else {
      callback(null, tenantUaaConfigResolver(uaaOptions, requestHost, tenantHostPattern));
    }
  }
}

function tenantUaaConfigResolver(uaaOptions, requestHost, tenantHostPattern) {
  var tenant = retrieveTenantFromURL (requestHost, tenantHostPattern);
  uaaOptions.tenant = tenant;
  var tenantUaaHost = (tenant ? tenant + '.' : '') + uaaOptions.uaadomain;
  uaaOptions.url = toUrlString(url.parse(uaaOptions.url), tenantUaaHost);

  tracer.debug('Tenant uaaUrl: request host "%s", TENANT_HOST_PATTERN: %s, uaa url "%s"', requestHost, tenantHostPattern, uaaOptions.url);

  return uaaOptions;
}

function toUrlString(parsedUaaUrl, tenantUaaHost) {
  return url.format({
    protocol: parsedUaaUrl.protocol,
    slashes: parsedUaaUrl.slashes,
    auth: parsedUaaUrl.auth,
    hostname: tenantUaaHost,
    port: parsedUaaUrl.port,
    pathname: parsedUaaUrl.pathname,
    search: parsedUaaUrl.search,
    hash: parsedUaaUrl.hash
  });
}

function retrieveTenantFromURL (approuterHost, tenantHostPattern) {
  approuterHost = approuterHost.split(':')[0]; // remove port
  var tenantHostPatternMatches = tenantHostPattern.exec(approuterHost);
  var tenant = tenantHostPatternMatches && tenantHostPatternMatches[1];
  return tenant;
}
